include/kernel.h

```diff
@@ -537,6 +539,17 @@ __syscall k_tid_t k_current_get(void);
  * released. It is the responsibility of the caller of this routine to ensure
  * all necessary cleanup is performed.
  *
+ * After k_thread_abort() returns, the thread is guaranteed not to be
+ * running or to become runnable anywhere on the system.  Normally
+ * this is done via blocking the caller (in the same manner as
+ * k_thread_join()), but in interrupt context on SMP systems the
+ * implementation is required to spin for threads that are running on
+ * other CPUs.  Note that as specified, this means that on SMP
```

Comment:
As a note for future development here, there is a way to deal with this, it's just not great.

(Not something to do in this PR though!)

(Namely, if you're in interrupt context and you are spinning waiting for a thread to be aborted, check to see if the thread you interrupted interrupted thread should be aborted, and if so handle it yourself. There are a bunch of ordering considerations, but that's the general gist of it.)

Reply:
Yeah, in the general case that becomes a full graph search for cycles .   Real devices (not ones running Zephyr, but then you never know...) exist with hundreds of CPUs.  You'd need to be willing to search that graph while holding a global scheduler lock literally in the middle of a context switch.  In an "RTOS". :)

FWIW: I went digging to see what Linux does here, and the answer is actually it doesn't.  There's no accessible API I can find to kill a process from an interrupt at all.  You kill processes by sending them signals so the abort is always done from their own kernel stack.  So even having documented the deadlock, I think we're ahead of the game here.

Reply:
>Yeah, in the general case that becomes a full graph search for cycles .

Hmm. Perhaps we're thinking of something different here (or you've identified something I'm missing, which is entirely likely).

The check for "if the thread you interrupted should be aborted" is constant-time, assuming you check the thread's flag. No searching the graph with BigLock.

At a conceptual level:

1. Every time you go to abort a thread, you ensure that your target thread is marked as aborting before you drop into `join` (...which we already do).
2. If you're spinning in a join in interrupt context, and the interrupted thread is marked as aborting, handle that abort.

(The key here is that there _is_ an implicit graph search, but it's incremental and amortized across all of the `k_thread_abort` calls. See, the only way to _create_ the graph of aborting threads is with a call to `k_thread_abort` for each edge... and each node has flags. If you squint a lot this turns into graph labelling...)

(There are still problems here with e.g. aborting a thread that's holding a mutex I am well aware. But this at least solves the cycle-of-aborting-threads issue.)

Reply:
Ah, got it.  I was talking about being able to detect the cycle of parallel aborts and return an error to a user who tries.

But you're right, being able to detect a newly aborted _current from within that spin loop would have the effect of breaking the cycle.  That's clever, I'll see what needs to be done, but yeah, maybe in a separate PR.